src/day8/re_frame/test.cljc

```diff
@@ -1,66 +1,276 @@
+  "Run `body` as an async re-frame test. The async nature means you'll need to
+  use `wait-for` any time you want to make any assertions that should be true
+  *after* an event has been handled.
+
+  This macro will automatically clean up any changes to re-frame state made
+  within the test body, as per `with-temp-re-frame-state` (except that the way
+  it's done here *does* work for async tests, whereas that macro used by itself
+  doesn't)."
+  [& body]
+  `(run-test-async* (fn [] ~@body)))
+
+
+(defn- as-callback-pred
+  "Interprets the acceptable input values for `wait-for`'s `ok-ids` and
+  `failure-ids` params to produce a predicate function on an event.  See
+  `wait-for` for details."
+  [callback-pred]
+  (when callback-pred
+    (cond (or (set? callback-pred)
+              (vector? callback-pred)) (fn [event]
+                                         (some (fn [pred] (pred event))
+                                               (map as-callback-pred (seq callback-pred))))
+          (fn? callback-pred)          callback-pred
+          (keyword? callback-pred)     (fn [[event-id _]]
+                                         (= callback-pred event-id))
+          :else                        (throw
+                                        (ex-info (str (pr-str callback-pred)
+                                                      " isn't an event predicate")
+                                                 {:callback-pred callback-pred})))))
+
+
+(defn wait-for* [ok-ids failure-ids callback]
+  (let [{:keys [done] :as test-context} (update *test-context* :wait-for-depth inc)]
+    (swap! (:max-wait-for-depth test-context) inc)
+
+    (let [ok-pred   (as-callback-pred ok-ids)
+          fail-pred (as-callback-pred failure-ids)
+          cb-id     (gensym "wait-for-cb-fn")]
+      (rf/add-post-event-callback cb-id (fn [event _]
+                                          (cond (and fail-pred
+                                                     (not (test/is (not (fail-pred event))
+                                                                   "Received failure event")))
+                                                (do
+                                                  (rf/remove-post-event-callback cb-id)
+                                                  (reset! (:now-waiting-for test-context) nil)
+                                                  (done))
+
+                                                (ok-pred event)
+                                                (do
+                                                  (rf/remove-post-event-callback cb-id)
+                                                  (reset! (:now-waiting-for test-context) nil)
+                                                  (binding [*test-context* test-context]
+                                                    (callback event))
+                                                  (when (= (:wait-for-depth test-context)
```

Comment:
It would be good to explain the implementation/internal details of wait-for-depth a little bit further. I get the gist of what you're doing, and the intent, but don't fully understand it.